Add settings for Stork input and output options

lioman opened the pull request:
Stork index creation can be configured via [additional options](https://stork-search.net/docs/config-ref).
To allow all of them add SEARCH_ADDITIONAL_OPTIONS settings.
Content will be added to stork.toml in input block.

# Pull Request Checklist

Resolves: #4  <!-- Only if related issue *already* exists — otherwise remove this line -->

<!-- This is just a reminder about the most common mistakes. Please make sure that you tick all *appropriate* boxes. Also, please read our [contribution guide](https://docs.getpelican.com/en/latest/contribute.html#contributing-code) at least once — it will save you unnecessary review cycles! -->

- [x] Conformed to **code style guidelines** by running appropriate linting tools
- [x] Updated **documentation** for changed code

<!-- If you have *any* questions to *any* of the points above, just **submit and ask**! This checklist is here to *help* you, not to deter you from contributing! -->

lioman committed:
#4: add additional options

stork index creation can be configured via additional options.
To allow all of them add SEARCH_ADDITIONAL_OPTIONS settings.
Content will be added to stork.toml in input block.

lioman committed:
Fix documentation

lioman commented:
Do we need anything to get this merged? I don't want to pin this dependency to git but use a normal version instead

justinmayer commented:
Hey Lioman. Anything that adds/changes settings has to be considered a bit more carefully than other contributions, since changing them later can cause problems for people using previous setting names. I am completely swamped at the moment, so all I can promise is that I will review this in depth as soon as I possibly can.

lioman commented:
Sure! No problem, thanks for answering. I added as additional parameter. So if you do not add it, nothing changes.

lioman commented:
Sorry to ping you again. @justinmayer Is it possible, that you review this? I hate long pending tasks

Kristinita commented:
**Type: Request** :mailbox_with_mail:

### 1. Request

@lioman , could you make that Pelican users add options in the traditional Pelican format, which is used in other plugins?

Example:

```python
SEARCH_INPUT_OPTIONS = {
	"displayed_results_count": 100
}
```

### 2. Argumentation

[**I store my Pelican settings in `.yaml`**](https://github.com/Kristinita/KristinitaPelican/blob/master/pelican-config/pelican_plugins_settings.yaml), not `.py` files. YAML is a specifically configuration file format, and [**I don’t know**](https://stackoverflow.com/q/49901577/5951529) how to parse the configuration in Python files directly.

If the settings after your change could be entered in the traditional format, YAML users could add them like this:

```yaml
SEARCH_INPUT_OPTIONS:
  displayed_results_count: 100
```

I don’t know, how to convert [**your chosen option format**](https://github.com/pelican-plugins/search/pull/5/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R68) to YAML.

Thanks.

lioman commented:
@Kristinita I can, but this would be way more complicated, as the settings need to be parsed and translated to a toml format. I chose the current because it is the simplest approach. Basically, it is just a string the gets directly injected and for your yaml setting this should be possible too. Have not tested it, but this should be valid
```yaml
SEARCH_INPUT_OPTIONS: |
    stemming = "German"
    url_prefix = "https://example.com"
```

Kristinita commented:
**Type: Another request** :mailbox_with_mail:

Thanks, it works for me!

But when I try to change the value of the option `displayed_results_count`, I get an error:

```markdown
Exception: Search plugin reported Error: Couldn't read the configuration file: Cannot parse config as TOML. Stork recieved error: unknown field `displayed_results_count`, expected one of `surrounding_word_count`, `base_directory`, `url_prefix`, `title_boost`, `stemming`,
`html_selector`, `exclude_html_selector`, `frontmatter_handling`, `files`, `break_on_file_error`, `srt_config`, `minimum_indexed_substring_length`, `minimum_index_ideographic_substring_length` for key `input` at line 772 column 1`
```

It seems that due to your solution, it’s solely possible to change the values for the so-called Stork “Input Options”. But [**Stork configuration**](https://stork-search.net/docs/config-ref) isn’t solely “Input Options”.

@lioman , what are your thoughts on making it possible to change the values of another Stork options as well for Pelican users?

Thanks.

lioman commented:
I discovered the output options later and then I want to wait till this PR is merged and implement the output options in the same way.

But apparently @justinmayer hasn't found the time yet to review it. I can change it now so that in- and output options are configurable but I want to know if we can do it this way.

Kristinita commented:
**Type: Another request** :mailbox_with_mail:

@lioman, You also use `stemming = "German"` in your example, but [**according to Stork `stemming` documentation**](https://stork-search.net/docs/languages) I think the `stemming` value should be generated completely automatically for the Pelican based on user Pelican settings. Example of generated `search.toml`:

```toml
[input]
base_directory = "D:/Kristinita/output"
html_selector = "section"
# [INFO] If “DEFAULT_LANG = 'ru'” in Pelican settings
# https://docs.getpelican.com/en/latest/settings.html#translations
stemming = "Russian"

# [INFO] The article on default language for the project,
# no “stemming_override”
[[input.files]]
path = "Статья-на-русском-языке.html"
url = "/Статья-на-русском-языке.html"
title = "Статья на русском языке"

# [INFO] “Lang: en” metadata in Pelican article or page:
# https://docs.getpelican.com/en/latest/content.html#file-metadata
[[input.files]]
path = "Article-in-English.html"
url = "/Article-in-English.html"
title = "Article in English"
stemming_override = "English"

```

Thanks.

justinmayer commented:
Really sorry for the delay, Lioman. I'll be on vacation until the end of September but will do my utmost to review it upon my return. Many thanks for all your patience! 😅

lioman commented:
> Really sorry for the delay, Lioman. I'll be on vacation until the end of September but will do my utmost to review it upon my return. Many thanks for all your patience! 😅

No problem - have a nice vacation

lioman commented:
> **Type: Another request** 📬
> 
> @lioman, You also use `stemming = "German"` in your example, but [**according to Stork `stemming` documentation**](https://stork-search.net/docs/languages) I think the `stemming` value should be generated completely automatically for the Pelican based on user Pelican settings.…

I don't get what you say here. but perhaps this is for another issue to automatically detect the language based on file language. This would be another issue/pr and a different discussion.
This should not change anything here as this one is just for setting anything, that is not already covered.

I'm even willed to implement it. As this would be nicer, to have this automatically covered and no need to think about it

paulocoutinhox commented:
It will be merged?

lioman committed:
Merge branch 'main' into main

lioman commented:
@justinmayer please review it - it is open so long now

nandac commented:
@lioman @justinmayer I would love to see this feature merged and released. Is there any way I may help with this?

lioman commented:
> @lioman @justinmayer I would love to see this feature merged and released. Is there any way I may help with this?

Me too, waiting a long time for it

justinmayer commented:
Hey folks. I'm aware, and I'm doing the best I can to allocate time to review this PR this month. Thanks for your patience.

lioman commented:
Thanks @justinmayer ! If I can do anything to help, just call out.

lioman committed:
Merge branch 'main' into main

lioman commented:
@justinmayer I just resolved the conflicts - hope you find some time to review this

lioman commented:
@justinmayer sorry to ping you again. I just want to complete this issue. Do I need to change something or is it fine like it is?

justinmayer commented:
First, thanks to @lioman for taking the time to implement and submit this pull request.

My main concern with the implementation in its current form is that I originally added two of the more important "input" options as separate settings. It seems awkward to add the rest of the Stork input options in a separate monolithic setting, with the original two dangling off by themselves. I imagine it would be better to either add all the settings as separate (which seems hard to maintain) or to move the original two separate input settings into a new monolithic setting. The latter seems like the far more sensible approach to me, of course with a deprecation period that maintains backwards compatibility for a suitable length of time.

Perhaps we could start off with new `STORK_INPUT_OPTIONS`, `STORK_OUTPUT_OPTIONS`, etc., since these settings are no longer tied to the Search plugin so much as they are to Stork itself. I think they would probably be best as key/value Python `dict` objects that are then unpacked into the TOML form that Stork expects.

Thoughts?

lioman commented:
I deliberately chose to handle those input options differently, as the other two should come from the rest of config. But that was due to my simple approach of using just a plain string. With a python object I think we have more power here and I will implement it with a sensible default, so that nothing changes for the ones without changes.
I suggest, to introduce `tomli-w` as a dependency. (Just learned that `tomllib`/`tomli` can only read toml)
Is that fine @justinmayer ?

justinmayer commented:
Sure, adding a dependency to more easily write TOML should be fine. As far as available options, in addition to https://github.com/hukkin/tomli-w, there is also [TOMLkit](https://github.com/sdispater/tomlkit). I've never used either, so feel free to use whichever seems like the best fit.

lioman committed:
add control over input and output options #4

- refactor
- add tests
- adapt documentation
- use python objects as configuration
- render configuration with rtoml

lioman committed:
Merge branch 'pelican-plugins:main' into main

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

lioman commented:
@justinmayer I have rewritten the hole stork settings generation with support for input and output options.
Additionally, I added some tests, so workflow needs an approval first. Please be so kind and review it soon.

justinmayer commented on .github/workflows/main.yml:
```diff
@@ -28,7 +28,9 @@ jobs:
-        run: poetry run invoke lint --diff
+        run: poetry run invoke lint
+      - name: Run tests
+        run: poetry run invoke tests
```
Why was `--diff` removed here?

Also, please add a newline between these two steps (linters and tests).

justinmayer commented on .gitignore:
```diff
@@ -1 +1,5 @@
+__pycache__
+coverage
+.coverage
+.vscode
```
My perspective is that most `.gitignore` directives belong on individual developers’ workstations in their global `~/.gitignore` files, leaving only the project-specific directives in project repositories like this one. Even though this project isn't using Coverage yet, I suppose the two Coverage-related files could be useful in the future. My suggestion is to change this file to the following contents:
```
.coverage
coverage
poetry.lock
```

justinmayer commented on README.md:
```diff
@@ -5,7 +5,6 @@
 
 This plugin generates an index for searching content on a Pelican-powered site.
 
-
```
The extra newlines removed from the README in this PR were intentionally added to improve readability when viewing/editing Markdown as raw, unrendered text (which is the way I interact with Markdown files 99% of the time).

These kinds of stylistic, preference-related changes are better discussed in a separate context, so let's revert these newline deletions.

justinmayer commented on README.md:
```diff
@@ -36,18 +34,27 @@ Once Stork has been successfully installed and tested, this plugin can be instal
 
-If you are using Pelican 4.5+ with namespace plugins and don’t have a `PLUGINS` setting defined in your configuration, then the Search plugin should be auto-discovered with no further action required. If, on the other hand, you _do_ have a `PLUGINS` setting defined (because you also use legacy plugins or because you want to be able to selectively disable installed plugins), then you must manually ...
 
+If you are using Pelican 4.5+ with namespace plugins and don’t have a `PLUGINS` setting defined in your configuration, then the Search plugin should be auto-discovered with no further action required. If, on the other hand, you _do_ have a `PLUGINS` setting defined (because you also use legacy plugins or because you want to be able to selectively disable installed plugins), then you must manually ...
```
I can't tell what change is being made here, so let's unless that can be made clear, let's revert it.

justinmayer commented on README.md:
```diff
@@ -36,18 +34,27 @@ Once Stork has been successfully installed and tested, this plugin can be instal
 
-In addition to plain-text files, Stork can recognize and index HTML and Markdown-formatted content. The default behavior of this plugin is to index generated HTML files, since Stork is good at extracting content from tags, scripts, and styles. But that mode may require a slight theme modification that isn’t necessary when indexing Markdown source (see `SEARCH_HTML_SELECTOR` setting below). That s ...
+In addition to plain-text files, Stork can recognize and index HTML and Markdown-formatted content. The default behavior of this plugin is to index generated HTML files, since Stork is good at extracting content from tags, scripts, and styles. But that mode may require a slight theme modification that isn’t necessary when indexing Markdown source (see HTML selector setting below). That said, inde ...
+Any other setting then the output path will toggle the plugin to switch to "source" mode.
```
Let's add a newline above this sentence. Otherwise Markdown will render it as appended it to the previous paragraph.

justinmayer commented on README.md:
````diff
@@ -36,18 +34,27 @@ Once Stork has been successfully installed and tested, this plugin can be instal
+}
+```
+
+#### Stork HTML selector
````
Capital "S" `Selector`, just for consistency with other headings.

justinmayer commented on README.md:
```diff
@@ -72,6 +79,38 @@ To use the default `main` selector, in each of your theme’s relevant template
 
+**Example**:
+
+To set it to a different selector, you can set it like this (e.g. primary).
```
Suggested modification to language/formatting:

> To set it to a different selector (for example, `primary`), you can set it like this:

justinmayer commented on README.md:
````diff
@@ -72,6 +79,38 @@ To use the default `main` selector, in each of your theme’s relevant template
+}
+```
+
+Additional [Input Options](https://stork-search.net/docs/config-ref) can be added here as a dict.
````
dict → `dict`

justinmayer commented on README.md:
```diff
@@ -72,6 +79,38 @@ To use the default `main` selector, in each of your theme’s relevant template
+### `STORK_OUTPUT_OPTIONS = ""`
+
+[Output Options](https://stork-search.net/docs/config-ref) can be configured as a dict.
+Keep in mind that keys are case sensitive. They mus be lower case.
```
dict → `dict`

case sensitive… → case-sensitive and must be lower-case.

justinmayer commented on README.md:
```diff
@@ -191,4 +232,4 @@ To start contributing to this plugin, review the [Contributing to Pelican][] doc
 [Pelican plugins documentation]: https://docs.getpelican.com/en/latest/plugins.html
 [existing issues]: https://github.com/pelican-plugins/search/issues
-[Contributing to Pelican]: https://docs.getpelican.com/en/latest/contribute.html
+[contributing to pelican]: https://docs.getpelican.com/en/latest/contribute.html
```
Unless there's a clear reason for it, let's revert this change.

lioman replied:
The reason behind this was a bug in prettier.io which is resolved in [2.8.2](https://github.com/prettier/prettier/blob/main/CHANGELOG.md#282) . Will change it back to the original one.

justinmayer reviewed (changes requested):
Nice work, Lioman! I proposed some minor suggested changes. I haven't tested the code itself but assume it works well. Let me know when everything is ready for another look. 💫

lioman committed:
minor changes after code review

lioman committed:
fix missed changes

lioman committed:
add empty lines for better readability

lioman commented:
Hi @justinmayer I changed the documentation according to your remarks. If we good to go for a merge I suggest to squash the commits together as there are many - or we decline this one and I open one with this branch: https://github.com/lioman/search/tree/4-add-control-over-input-output-options

justinmayer committed:
Update README.md

justinmayer committed:
Update .gitignore

justinmayer committed:
Merge branch 'main' into main

justinmayer reviewed (approved):
@lioman: Nice work. Thank you for addressing those minor points, some of which I tweaked further myself to save you the trouble of more back-and-forth.

Would you please add a `RELEASE.md` file in the project root, with release type of `minor`? You can see the following commit to see an example `RELEASE.md` file that was used to publish the last release: https://github.com/pelican-plugins/search/commit/cb1cb66898462d6b08af7ed8ce207ec9992ae641

justinmayer committed:
Fix some text

justinmayer committed:
Fix spurious characters introduced in v1.0.2

justinmayer committed:
Prepare release [skip ci]

justinmayer commented:
@lioman: I have taken care of adding the `RELEASE.md` file and am about to merge this PR, which should trigger a new release.

As a suggestion for the future, I highly recommend doing all work on a feature/topic branch instead of the default `main` branch, even in the context of your own fork. Doing so really does help make things easier for maintainers. In this case, the release automation logic got confused by the `main` branch designation (even though it wasn't *this* repository's `main` branch) and tried to issue a release upon a push event, even though the PR hadn't been merged yet. Using a feature/topic branch would have prevented that problem.

Thank you for all your work on — and patience with — this pull request. This is a really nice enhancement to the plugin. Much appreciated! ✨
